cdap-ui/server/url-helper.js

```diff
@@ -14,28 +14,57 @@
  * the License.
 */
 
+import memoize from 'lodash/memoize';
+
 export const REQUEST_ORIGIN_ROUTER = 'ROUTER';
 export const REQUEST_ORIGIN_MARKET = 'MARKET';
 
+function getRouterHost(cdapConfig) {
+  const routerhost = cdapConfig['router.server.address'],
+    routerport =
+      cdapConfig['ssl.external.enabled'] === 'true'
+        ? cdapConfig['router.ssl.server.port']
+        : cdapConfig['router.server.port'],
+    routerprotocol = cdapConfig['ssl.external.enabled'] === 'true' ? 'https' : 'http';
+  return `${routerprotocol}://${routerhost}:${routerport}`;
+}
+
+function extractMarketUrls(cdapConfig) {
+  if (!cdapConfig) {
+    return [];
+  }
+  const defaultMarketUrl = cdapConfig['market.base.url'];
+  if (!cdapConfig['market.base.urls']) {
+    return [defaultMarketUrl];
+  }
+  const marketUrls = cdapConfig['market.base.urls'].split('+');
+  const filteredMarketUrls = marketUrls.filter(element => element !== defaultMarketUrl); 
+  // Make sure the default CDAP market is the first market.
+  filteredMarketUrls.splice(0, 0, defaultMarketUrl);
+  return filteredMarketUrls;
+}
+
+export const getMarketUrls = memoize(extractMarketUrls);
+
+export function isVerifiedMarketHost(cdapConfig, url) {
+  return !!getMarketUrls(cdapConfig).find(element => url.startsWith(element));
+}
+
 export function constructUrl(cdapConfig, path, origin = REQUEST_ORIGIN_ROUTER) {
   if (!cdapConfig) {
     return null;
   }
   path = path && path[0] === '/' ? path.slice(1) : path;
   if (origin === REQUEST_ORIGIN_MARKET) {
-    return `${cdapConfig['market.base.url']}/${path}`;
+    return isVerifiedMarketHost(cdapConfig, path) ? path : 'invalid-market-url';
```

Comment:
nit: Would we have to verify the host here? Shouldn't this be already done on any market url we get before constructing the url? If we return `invalid-market-url` I believe express server will still make the call right?